test/decorate_all_test.exs

```diff
@@ -0,0 +1,60 @@
+defmodule DecoratorTest.Fixture.DecorateAllDecorator do
+  use Decorator.Define, [some_decorator: 1]
+
+  def some_decorator(message, body, _context) do
+    quote do
+      send self(), unquote(message)
+      unquote(body)
+    end
+  end
+end
+
+defmodule DecoratorTest.Fixture.DecorateAllTestModule do
+  use DecoratorTest.Fixture.DecorateAllDecorator
+
+  @decorate_all some_decorator(:we_did_it)
+
+  def fun_one do
+    :ok
+  end
+
+  def fun_two do
+    :ok
+  end
+
+  def fun_two(_arg1, something \\ :default)
```

Comment:
Something is fishy here. It emits warning about a missing header, although you put one in; and also about an unused variable (`something`) which is not unused:

```
warning: definitions with multiple clauses and default values require a header. Instead of:

    def foo(:first_clause, b \\ :default) do ... end
    def foo(:second_clause, b) do ... end

one should write:

    def foo(a, b \\ :default)
    def foo(:first_clause, b) do ... end
    def foo(:second_clause, b) do ... end

def fun_two/2 has multiple clauses and defines defaults in one or more clauses
  test/decorate_all_test.exs:12

warning: this clause cannot match because a previous clause at line 12 always matches
  test/decorate_all_test.exs:12

warning: variable "something" is unused
  test/decorate_all_test.exs:25
```

I think that the header is transformed into a normal function and the original function is not executed, can that be? When I do a `raise` into the definition of `fun_two/2`, the tests still pass.

Reply:
Hmm yes it seems that function headers are decorated resulting in a catch-all function clause with an empty body. Any good ideas on how to determine whether a particular function is a header and should be ignored? 🤔